### DISCUSSION TITLE
Analyzer cannot read sample actisense-w2k-n2k.raw

### DISCUSSION THREAD

**Discussion by vidax** [canboat/canboat#371] [Q&A] [answered]
I am trying to analyze the actisense-w2k-n2k.raw from the sample folder, and I get this:
analyzer -json < actisense-w2k-n2k.raw
{"version":"4.9.2","units":"std","showLookupValues":false}
ERROR 2023-02-09T08:33:26.812Z [analyzer] Unknown message format

The ASCII format from the file is: 
A000057.055 09FF7 0FF00 3F9FDCFFFFFFFFFF

Am I missing a step in the process?

**Comment by aldas** [comment]
`analyzer` only recognizes canboat own raw format. You need to first convert actisense ascii to canboat and then analyzer can decode it to fields.

NB: `actisense-serial` does not support Actisense ASCII format.  W2k binary ascii format is almost same as NGT1 format (that canboat `actisense-serial` supports but some constanst are little bit different so `actisense-serial` can not read it at the moment.



This is  Actisense ASCII  format (I got it from Actisense support)
```
	// Ahhmmss.ddd <SS><DD><P> <PPPPP> b0b1b2b3b4b5b6b7.....bn<CR><LF>
	// A = message is N2K or J1939 message
	// 173321.107 - time 17:33:21.107
	// <SS> - source address
	// <DD> - destination address
	// <P> - priority
	// <PPPPP> - PGN number
	// b0b1b2b3b4b5b6b7.....bn - data payload in hex. NB: ISO TP payload could be up to 1786 bytes
	//
	// Example: `A173321.107 23FF7 1F513 012F3070002F30709F\n`
	//                      1     2     3                  4
```

Canboat format is:
```
2022-10-11T11:47:22Z,6,60928,16,255,8,99,ad,22,22,00,a0,64,c0
1                   ,2,3    ,4 ,5  ,6,7...

1 = time RFC3339Nano
2 = priority
3 = PGN
4 = source
5 = destination
6 = data length in bytes
7 = data as hex, each byte separated by comma
```

so you probably can write small script to turn Actisense ASCII to raw canboat format

**Reply by aldas** [reply]
I am not proficient in C but I have multiple Actisense implementations in Go

N2K binary: https://github.com/aldas/go-nmea-client/blob/aa379a8611586f8e2dc790fe94f56ef0adc9c897/actisense/binaryreader.go#L246
it is very similar to NGT binary but still different.

Actisense ASCII: https://github.com/aldas/go-nmea-client/blob/aa379a8611586f8e2dc790fe94f56ef0adc9c897/actisense/n2kasciidevice.go#L106

W2K RAW Ascii: https://github.com/aldas/go-nmea-client/blob/aa379a8611586f8e2dc790fe94f56ef0adc9c897/actisense/rawasciidevice.go#L99

Test data/examples could be found here: https://github.com/aldas/go-nmea-client/tree/main/actisense/testdata

**Comment by vidax** [comment]
Thank you, Martti! And nice to meet you, happy new year and all.
--> At first all we got from the actisense API was the .EBL binary logs. I
did not find any library that would open them, so we had a talk with
Actisense, and they proposed to get us a script to convert those EBL to
ASCII as in the sample file.

I will look into the file you sent, and see if I or some of my devs are
competent enough to work from this. (I doubt it)
Would you be open to being contracted to create this piece of the script? I
would be happy to jump on a call, I am sure you will understand what we are
trying to do right away.

Let me know.

David Culot

**Reply by aldas** [reply]
This needs Bash, probably WSL helps here

Create file `actisense-ascii-to-canboat.sh` with following contents:
```bash
#!/bin/bash

ACTISENSE_FILE=$1
if [[ ! -f "$ACTISENSE_FILE" ]]; then
  echo "$ACTISENSE_FILE file not found!"
  exit 1
fi

# Actisense ASCII format:
# Ahhmmss.ddd <SS><DD><P> <PPPPP> b0b1b2b3b4b5b6b7.....bn<CR><LF>
# A = message is N2K or J1939 message
# 173321.107 - time 17:33:21.107
# <SS> - source address
# <DD> - destination address
# <P> - priority
# <PPPPP> - PGN number
# b0b1b2b3b4b5b6b7.....bn - data payload in hex. NB: ISO TP payload could be up to 1786 bytes
#	//
# Example: `A173321.107 23FF7 1F513 012F3070002F30709F\n`

# Canboat format: 2022-11-15T09:24:06.654Z,6,60928,128,255,8,a4,07,33,e8,00,a0,50,c0
while IFS= read -r line
do
  if [[ ${#line} == 0 ]]; then
    continue
  fi
  # if value of $line starts with #, ignore it
  if [[ $line =~ ^#.* ]]; then
    continue
  fi
  HOUR=${line:1:2}
  MINUTE=${line:3:2}
  SEC=${line:5:6}
  SRC=$((16#${line:12:2}))
  DEST=$((16#${line:14:2}))
  PRIO=${line:16:1}
  PGN=$((16#${line:18:5}))
  DATA_RAW=${line##* }
  LENGTH=$((${#DATA_RAW} / 2))
  DATA=$(echo $DATA_RAW | sed 's/\(.\{2\}\)/\1,/g') # split DATA_RAW into 2 characters separated by comma


  echo "2023-02-09T${HOUR}:${MINUTE}:${SEC},${PRIO},${PGN},${SRC},${DEST},${LENGTH},${DATA}"
done < "$ACTISENSE_FILE"
echo ""
```

(optional) make that file executable `chmod +x actisense-ascii-to-canboat.sh`

Assuming you have file `actisense-w2k1-n2k-ascii.txt` formatted in Actisense ASCII format

```
A173321.107 23FF7 1F513 012F3070002F30709F
A000057.131 09FF2 1FE12 FF7FFF7FFBFFFCFFFF7F1300
```

now convert it to canboat format  and pipe it to canboat analyzer

```bash
./actisense-ascii-to-canboat.sh actisense-w2k1-n2k-ascii.txt | analyzer -json
```

and you should get output as
```
{"version":"4.9.1","units":"std","showLookupValues":false}
INFO 2023-02-09T10:41:47.125Z [analyzer] Detected normal format with all frames on one line
{"timestamp":"2022-11-15T17:33:21.107","prio":7,"src":35,"dst":255,"pgn":128275,"description":"Distance Log","fields":{"Date":"2002.12.12","Time":"21:54:15.7872","Log":10448944}}
{"timestamp":"2022-11-15T00:00:57.131","prio":2,"src":9,"dst":255,"pgn":130578,"description":"Vessel Speed Components","fields":{"Longitudinal Speed, Ground-referenced":-0.005,"Transverse Speed, Ground-referenced":-0.004,"Stern Speed, Ground-referenced":0.019}}
```

**Reply by aldas** [reply]
it might be that EBL is same format as raw NGT1 outputs. Here is someones project https://github.com/TwoCanPlugIn/ActisensePlugin/blob/master/src/actisense_ebl.cpp#L66 reading EBL files. It seems very similar to NGT1. Parsing is done here https://github.com/TwoCanPlugIn/ActisensePlugin/blob/953afc1eb506da464ec46f7eff9db7665fb29ee1/src/actisense_device.cpp#L236

Canboats `actisense-serial` command can read data from files also. So you might want to try `actisense-serial ~/mylog.ebl | analyzer -json`

**Comment by vidax** [comment]


**Reply by aldas** [reply]
I think Canboat is MVP (most valuable player) here and I propose asking @keesverruijt to enable Github Sponsors  for Canboat so you could choose some option there to show appreciation for canboat.

**Reply by keesverruijt** [reply]
> I think Canboat is MVP (most valuable player) here and I propose asking @keesverruijt to enable Github Sponsors for Canboat so you could choose some option there to show appreciation for canboat.

Thanks that is a good idea.

**Comment by keesverruijt** [comment]
@vidax, 

I added the w2k sample because I had a very short opportunity to sample some output coming from the new Actisense W2K-1. I did not have the time to adapt `analyzer` so that it can read these files.

Also, please cut your signature from email replies to GitHub, now they are in this discussion. I will remove them.

@aldas, thanks for your help. Just know that this statement is wrong:
> `analyzer` only recognizes canboat own raw format. You need to first convert actisense ascii to canboat and then analyzer can decode it to fields.

In fact `analyzer` knows a number of formats -- its own, one Airmar, one Chetco, two Garmin and one YachtDevices format. I will add this W2K format as it is easy to do.

For some reason :-) people find it hard to write C -- when it is the (grand)mother of most procedural languages, including Go... in the past I also allowed `candump2analyzer` converter to be added. This should probably be folded into analyzer as well.

**Reply by aldas** [reply]
well, I am being modest here :) I write occasionally stuff for microcontrollers (ESP32 etc) so C is not totally foreign to me but these are my hobby stuff where  memory leaks and stuff like that is OK or to be expected :)

**Comment by keesverruijt** [comment]
I opened issue #372.

**Comment by keesverruijt** [comment, marked as answer]
Issue #372 has been fixed and uploaded, and it's even released officially as v4.10.0.
